wgpu-core/src/command/ray_tracing.rs

```diff
@@ -0,0 +1,1223 @@
+        #[cfg(feature = "trace")]
+        if let Some(ref mut list) = cmd_buf.data.lock().as_mut().unwrap().commands {
+            list.push(trace::Command::BuildAccelerationStructuresUnsafeTlas {
+                blas: trace_blas.clone(),
+                tlas: trace_tlas.clone(),
+            });
+            if !trace_tlas.is_empty() {
+                log::warn!("a trace of command_encoder_build_acceleration_structures_unsafe_tlas containing a tlas build is not replayable!");
+            }
+        }
+
+        #[cfg(feature = "trace")]
+        let blas_iter = trace_blas.iter().map(|blas_entry| {
+            let geometries = match &blas_entry.geometries {
+                TraceBlasGeometries::TriangleGeometries(triangle_geometries) => {
+                    let iter = triangle_geometries.iter().map(|tg| BlasTriangleGeometry {
+                        size: &tg.size,
+                        vertex_buffer: tg.vertex_buffer,
+                        index_buffer: tg.index_buffer,
+                        transform_buffer: tg.transform_buffer,
+                        first_vertex: tg.first_vertex,
+                        vertex_stride: tg.vertex_stride,
+                        index_buffer_offset: tg.index_buffer_offset,
+                        transform_buffer_offset: tg.transform_buffer_offset,
+                    });
+                    BlasGeometries::TriangleGeometries(Box::new(iter))
+                }
+            };
+            BlasBuildEntry {
+                blas_id: blas_entry.blas_id,
+                geometries,
+            }
+        });
+
+        #[cfg(feature = "trace")]
+        let tlas_iter = trace_tlas.iter();
+
+        let mut input_barriers = Vec::<hal::BufferBarrier<dyn hal::DynBuffer>>::new();
+        let mut buf_storage = BufferStorage::new();
+
+        let mut scratch_buffer_blas_size = 0;
+        let mut blas_storage = BlasStorage::new();
+        let mut cmd_buf_data = cmd_buf.data.lock();
+        let cmd_buf_data = cmd_buf_data.as_mut().unwrap();
+
+        iter_blas(
+            blas_iter,
+            cmd_buf_data,
+            build_command_index,
+            &buffer_guard,
+            &blas_guard,
+            &mut buf_storage,
+        )?;
+
+        let snatch_guard = device.snatchable_lock.read();
+        iter_buffers(
+            &mut buf_storage,
+            &snatch_guard,
+            &mut input_barriers,
+            cmd_buf_data,
+            &buffer_guard,
+            &mut scratch_buffer_blas_size,
+            &mut blas_storage,
+        )?;
+
+        let mut scratch_buffer_tlas_size = 0;
+        let mut tlas_storage = Vec::<(
+            Arc<Tlas>,
+            hal::AccelerationStructureEntries<dyn hal::DynBuffer>,
+            u64,
+        )>::new();
+        let mut tlas_buf_storage = Vec::<(
+            Arc<Buffer>,
+            Option<PendingTransition<BufferUses>>,
+            TlasBuildEntry,
+        )>::new();
```

Comment:
I think it is worth defining structs for these, so the members get names and we don't get these complicated types.

Reply:
I've used types for some others, should those also be structs?

Reply:
If they're big like this, sure.

Reply:
They were bigger, clippy was complaining